### ISSUE TITLE
Add open rewrite to SpotBugs

### ISSUE THREAD

**Issue by JuditKnoll** [spotbugs/spotbugs#2761]
Creating this issue both as a reminder and a place to discuss it.

              Think we may benefit from adding open rewrite to our project.  See https://plugins.gradle.org/plugin/org.openrewrite.rewrite

_Originally posted by @hazendaz in https://github.com/spotbugs/spotbugs/issues/2753#issuecomment-1848589719_

**Comment by Pankraz76** [comment]
Nice to see this going, thanks.



Might want to consider the following recipes:

- https://docs.openrewrite.org/recipes/gradle/gradlebestpractices



As well as:


## Add `rewrite` support for `errorprone.refasterrules`

I’d like to propose integrating [Google’s Error Prone](https://github.com/google/error-prone) and its Picnic extension (demonstrated in [Automating Away Bugs with Error Prone | PlatformCon 2023](https://www.youtube.com/watch?v=6llnrUtVlrE)) to enable automated bug fixes via `rewrite` rules. This would complement Checkstyle’s static analysis capabilities by addressing semantic bugs rather than stylistic issues.

## Motivation
Error Prone’s `refaster`/`rewrite` rules can automatically:
- Fix common bug patterns (e.g., `String.equals()` misuse)
- Modernize code (e.g., JDK migration helpers)
- Enforce best practices (e.g., null-check improvements)

Real-world adoptions show tangible benefits:
- [Apache Kafka’s integration](https://github.com/apache/kafka/pull/20219)
- [Spotless’s implementation](https://github.com/diffplug/spotless/pull/2576)

## Proposal
1. Add support for `errorprone.refasterrules`-based rewrites
2. Implement with opt-in adoption (no breaking changes)
3. Include suppression mechanisms for API constraints

## Discussion Points
- Need consensus on:
  - Scope of auto-fixes
  - Preferred suppression strategy
  - Integration approach

## Next Steps
I’m happy to:
- Prepare a PoC demonstrating the value
- Collaborate on implementation strategy
- Address any concerns about compatibility



relates to:

- https://github.com/checkstyle/checkstyle/issues/16543
- https://github.com/checkstyle/checkstyle/issues/17487

**Comment by Pankraz76** [comment]
assuming this is not desired as rejected by @hazendaz.

Thanks for the conversation and feedback.

**Comment by Pankraz76** [comment]
still an incomplete refactoring:

- https://github.com/spotbugs/spotbugs/pull/2753/files#r2307322092
- https://docs.openrewrite.org/recipes/staticanalysis/modifierorder

perfect example why this is actually an good intention suggested.

**Comment by ThrawnCA** [comment]
I agree that it's not suitable to add. The way I see it, adding an unrelated second function limits the usefulness of SpotBugs, since it would then only be usable by the _intersection_ of "people who want to detect errors" and "people who want to automatically change possible errors."

It's not necessary to put a shovel and a fork at opposite ends of the same handle. Just have two tools.

**Comment by Pankraz76** [comment]
Might create a new issue then, as it's clear there is a "broken window" effect demanding better static code analysis and quality gates.

PMD is the next alternative, as it performs a similar function, though of course it doesn't cover everything, such as code migration.


> It's not necessary to put a shovel and a fork at opposite ends of the same handle. Just have two tools.


The point is not to have redundant tools. It's about using specialized tools that find specific issues others miss. If you want to identify all gaps, leaks, and issues before they bleed into production, this is necessary.

Every tool has its blind spots, pros, and cons.
